File: src/main/java/es/xan/servantv3/sensors/SensorVerticle.java
package es.xan.servantv3.sensors;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

import com.jcraft.jsch.JSchException;

import es.xan.servantv3.AbstractServantVerticle;
import es.xan.servantv3.Action;
import es.xan.servantv3.Constant;
import es.xan.servantv3.MessageBuilder;
import es.xan.servantv3.MessageBuilder.ReplyBuilder;
import es.xan.servantv3.SSHUtils;
import es.xan.servantv3.messages.Sensor;
import io.vertx.core.eventbus.Message;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import io.vertx.core.logging.Logger;
import io.vertx.core.logging.LoggerFactory;

public class SensorVerticle extends AbstractServantVerticle {
	
	private static final Logger LOGGER = LoggerFactory.getLogger(SensorVerticle.class);
	
	private String mHost;
	private String mLogin;
	private String mPassword;
	
	private Map<String, String> mSensors;
	
	public SensorVerticle() {
		super(Constant.SENSOR_VERTICLE);
		
		supportedActions(Actions.values());
	}
	
	public enum Actions implements Action {
		RESET_SENSOR(Sensor.class)
		;
		
		Class<?> beanClass;
		
		private Actions(Class<?> beanClass) {
			this.beanClass = beanClass;
		}

		@Override
		public Class<?> getPayloadClass() {
			return this.beanClass;
		}
		
	}
	
	public void start() {
		super.start();
		loadConfiguration(vertx.getOrCreateContext().config().getJsonObject("SensorVerticle"));
	}
	
	private void loadConfiguration(JsonObject config) {
		mHost = config.getString("server");
		mLogin = config.getString("usr");
		mPassword = config.getString("pws");
		
		loadSensors(config.getJsonArray("items"));
	}

	private void loadSensors(JsonArray array) {
		mSensors = new HashMap<>();
		
		for (int i=0; i < array.getList().size(); i++) {
			final JsonObject sensor = (JsonObject) array.getList().get(i);
			String name = sensor.getString("name");
			String command = sensor.getString("command");
			
			mSensors.put(name, command);
		}
	}

	public void reset_sensor(Sensor sensor, Message<Object> message) {
		LOGGER.info("Asking to reset sensor [{}]", sensor.getSensor());
		final String command = mSensors.get(sensor.getSensor());
		
		final ReplyBuilder builder = MessageBuilder.createReply();
		if (command == null) {
			LOGGER.warn("Sensor [{}] not found", sensor.getSensor());
			
			builder.setError();
			builder.setMessage("Allowed options " + mSensors.keySet());
			
			message.reply(builder.build());
		} else {
			
			boolean result = false;
			try {
				result = SSHUtils.runRemoteCommand(mHost, mLogin, mPassword, command);
			} catch (JSchException | IOException e) {
				LOGGER.warn(e.getMessage(), e);
			}
			
			if (result) builder.setOk();
			else builder.setError();
			
			message.reply(builder.build());
		}
	}



}

File: src/main/kotlin/es/xan/servantv3/sensors/SensorVerticle.kt
package es.xan.servantv3.sensors

import es.xan.servantv3.*
import es.xan.servantv3.messages.Sensor
import io.vertx.core.eventbus.Message
import io.vertx.core.json.JsonObject
import org.slf4j.LoggerFactory
import java.util.*
import kotlin.collections.HashMap

class SensorVerticle : AbstractServantVerticle(Constant.SENSOR_VERTICLE) {
	companion object {
        val LOG = LoggerFactory.getLogger(SensorVerticle::class.java.name)
    }
	
	init {
		supportedActions(Actions::class.java)
	}
	
	enum class Actions(val clazz : Class<*>? ) : Action {
		RESET_SENSOR(Sensor::class.java)
		;
	}
	
	val mHost		: String by lazy { config().getJsonObject("SensorVerticle").getString("server", "localhost") }
	val mLogin		: String by lazy { config().getJsonObject("SensorVerticle").getString("usr", "guest") }
	val mPassword	: String by lazy { config().getJsonObject("SensorVerticle").getString("pws", "guest") }
	val mSensors	: Map<String, String> by lazy {
		HashMap<String,String>().apply {
			config().getJsonObject("SensorVerticle").getJsonArray("items").forEach{ it ->
				val item = it as JsonObject
				val name = item.getString("name").replaceFirstChar { it.lowercase(Locale.getDefault()) }
				val command = item.getString("command")
				LOG.info("putting [$name]->[$command]")
				put(name, command)
			}
		}
	};

	/**
	 * Deprecated
	 */
	fun reset_sensor(sensor : Sensor , message : Message<Any> ) {
		LOG.info("Asking to reset sensor [{}]", sensor.sensor);
		
		val item = sensor.sensor.replaceFirstChar { it.lowercase(Locale.getDefault()) };
		
		val sensorsToReset = when (item) {
			"all" -> this.mSensors.keys;
			in mSensors.keys -> setOf(item);
			else -> {
				val builder = MessageBuilder.createReply()
				builder.setError();
				builder.setMessage("Options [$mSensors.keys]")
				message.reply(builder.build())
				
				return;
			}
		}
		
		val resetedWithError = sensorsToReset.filter { it ->
			val command = mSensors.get(it);
			try {
				SSHUtils.runRemoteCommand(mHost, mLogin, mPassword, command);
				false
			} catch (e: Throwable) {
				LOG.warn("Problems trying to manage ssh remote command [$command]", e);
				true
			}
		}
		
		val reply = MessageBuilder.createReply().apply {
			when {
				resetedWithError.isEmpty() -> setOk()
				else -> {
					setError()
					setMessage("Problems reseting items [$resetedWithError]")
				}
			}
		}
		
		message.reply(reply.build());
	}

}